crates/sui-graphql-rpc/schema.graphql

```diff
@@ -0,0 +1,840 @@
+# Copyright (c) Mysten Labs, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+# Open Questions
+# --------------
+# - Rules for nullabilit/non-nullability
+# - Pattern for multi-gets
+# - Encoding RPC errors
+# - `type` (vs `scalar`) for Move Types
+# - Pattern for upcasting and downcasting
+
+# Check Completeness Against Existing APIs
+# -------------
+# DONE Read
+# DONE - Objects
+# DONE - Transactions
+# DONE - Checkpoints
+# DONE - Events
+# DONE - ProtocolConfig
+# DONE - Chain Identifier
+# DONE Coin
+# DONE Extended
+# DONE Governance
+# DONE Indexer
+# DONE Move Utils
+# TODO Write
+
+# DONE Port designs from Notion
+# DONE - Display
+# DONE - SuiNS
+# DONE Data Consistency API
+# TODO Flesh out Connections
+# TODO - TotalTransactionBlocks for the PTB connection
+
+schema {
+  query: Query
+  subscription: Subscription
+}
+
+type Query {
+  # First four bytes of the network's genesis checkpoint digest
+  # (uniquely defines the network)
+  chainIdentifier: String!
+
+  # First and last checkpoint that the RPC can answer consistent
+  # queries for.
+  consistentRange: ConsistentRange
+
+  owner(address: SuiAddress!): Owner
+  object(address: SuiAddress!, version: U64): Object
+  address(address: SuiAddress!): Address
+
+  epoch(epochId: Int): Epoch
+  protocolConfig(protocolVersion: Int): ProtocolConfig
+
+  checkpoint(filter: CheckpointFilter): Checkpoint
+  checkpoints(order: Order!): CheckpointConnection
+
+  transactionBlock(digest: Digest!): TransactionBlock
+  transactionBlocks: TransactionBlockConnection
+
+  coinMetadata(coinType: TypeInst!): CoinMetadata
+
+  queryObjects(query: ObjectQuery): ObjectQueryConnection
+  queryEvents(filter: EventFilter, order: Order!): EventQueryConnection
+  queryTransactionBlocks(
+    filter: TransactionBlockFilter,
+    order: Order!,
+  ): TransactionBlockQueryConnection
+
+  resolveNameServiceAddress(name: String!): SuiAddress
+
+  networkMetrics: NetworkMetrics
+  moveCallMetrics: MoveCallMetrics
+  checkpointAddressMetrics(checkpoint: U64): AddressMetric
+  allEpochAddressMetrics(order: Order!): AddressMetricConnection
+}
+
+type Subscription {
+  events(filter: EventFilter): Event
+  transactions(filter: TransactionBlockFilter): TransactionBlock
+}
+
+# String containing 32B hex-encoded address
+scalar SuiAddress
+
+# String containing 32B Base58-encoded BLAKE2b256 digest
+scalar Digest
+
+# String representation of a 64 bit unsigned integer
+scalar U64
+
+# String representation of a 128 bit signed integer
+scalar I128
+
+# String representation of a 128 bit unsigned integer
+scalar U128
+
+# String containing Base58-encoded binary data.
+scalar Bytes
+
+scalar JSON
+
+# Fully qualified concrete type instantiation, with a flat
+# representation:
+#
+# ([package]::[module]::)?[name](<[param], ...>)?.
+scalar TypeInst
```

Comment:
I was thinking this would be helpful to document the API and to add data validation.  

Additionally for `TypeSignature` I. particular, the plan was for that not to be a flat string, but to have some structure as well (the main benefit of it being a scalar is that it's a recursive structure that gets fetched all in one go and it's opaque to GraphQL).

Reply:
Every scalar introduces client friction to set the SDK up, so we should use the absolute minimum amount possible while still conveying meaning. Validation can still happen on the server. Client validation depends on the consumer setting it up right at which point they can just do it manually anyway.

Reply:
Collapsed the scalars used down to just `SuiAddress`, `BigInt`, and `Bytes`, so hopefully this is a better balance.

Reply:
I think even `SuiAddress` could just be `ID` really.

Reply:
I was keeping these separate so that `ID` would be reserved for internal, opaque `ID`s, to make it clearer which primary key is used for cursors and which primary key is used by the network, but if that's not helpful, I'm happy to drop `SuiAddress`.

Reply:
I think it’s a good point that ID should be opaque and addresses are not. I’m not convinced that `string` isn’t sufficient here tho

Reply:
That makes sense to me, and I buy that, _although_ with a well connected graph we shouldn't really have too many floating addresses?

Reply:
It tends to show up quite often in `input` types (for filters).  For output types, it only shows up in a few places (`Owner`, `SharedInput`, `MoveAddress`, `MoveUID`, `ValidatorSet.reportRecords`)